src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+        // Name
+        if (timeZone.name.HasValue())
+        {
+            uint32_t name_size = static_cast<uint32_t>(strnlen(timeZone.name.Value().data(), 64));
+            ReturnErrorOnFailure(writer.PutString(TagName(), timeZone.name.Value().data(), name_size));
+            ChipLogProgress(Zcl, "%s %d", timeZone.name.Value().data(), name_size);
+        }
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        return CHIP_NO_ERROR;
+    }
+};
+
+static bool computeLocalTime(chip::EndpointId ep)
+{
+    DataModel::Nullable<uint64_t> utcTime, localTime;
+    int32_t timeZoneOffset = 0, dstOffset = 0;
+    UTCTime::Get(ep, utcTime);
+    if (utcTime.IsNull())
+    {
+        return false;
+    }
+    auto tz  = TimeSynchronizationServer::Instance().GetTimeZone().begin();
+    auto dst = TimeSynchronizationServer::Instance().GetDSTOffset().begin();
+    if (tz->validAt <= utcTime.Value())
+    {
+        timeZoneOffset = tz->offset;
+    }
+    if (dst->validStarting <= utcTime.Value() && dst->offset != 0 && !dst->validUntil.IsNull())
+    {
+        dstOffset = dst->offset;
+    }
+    localTime.SetNonNull(utcTime.Value() + static_cast<uint64_t>(timeZoneOffset) + static_cast<uint64_t>(dstOffset));
+    LocalTime::Set(ep, localTime);
+    return true;
+}
+
+class TimeSynchronizationAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the TimeSync cluster on all endpoints
+    TimeSynchronizationAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), TimeSynchronization::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+
+private:
+    CHIP_ERROR ReadTrustedTimeSource(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadDefaultNtp(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadTimeZone(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadDSTOffset(EndpointId endpoint, AttributeValueEncoder & aEncoder);
+};
+
+TimeSynchronizationAttrAccess gAttrAccess;
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadTrustedTimeSource(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto tts       = TimeSynchronizationServer::Instance().GetTrustedTimeSource();
+    if (!tts.IsNull())
+    {
+        err = aEncoder.Encode(tts.Value());
+    }
+    else
+    {
+        err = aEncoder.EncodeNull();
+    }
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationAttrAccess::ReadDefaultNtp(EndpointId endpoint, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    auto dntp      = TimeSynchronizationServer::Instance().GetDefaultNtp();
+    if (!dntp.IsNull())
+    {
+        const char * buf = reinterpret_cast<const char *>(dntp.Value().data());
+        err              = aEncoder.Encode(chip::CharSpan(buf, strnlen(buf, dntp.Value().size())));
```

Comment:
If we want a char string, `TimeSynchronizationServer::Instance().GetDefaultNtp()` should be returning a char string.  If we want a byte sequence, we should be doing a byte sequence.  There should be no reason for reinterpret_cast here.

Reply:
The issue is difference between the type used by the generated code and the type I use in the persistent storage implementation.

Reply:
My point is: why is there a difference?  We should just be plumbing the spec type through, imo.